File: src/python/core/src/dicebear/style_def/canvas.py
"""Read-only view over a style definition's ``canvas`` block."""

from __future__ import annotations

from typing import Any, cast

from .element import Element


class Canvas:
    """Read-only view over a style definition's ``canvas`` block.

    Exposes the drawing-area dimensions and the top-level element list.
    """

    def __init__(self, data: dict[str, Any]) -> None:
        self._data = data
        self._elements: list[Element] | None = None

    def width(self) -> int | float:
        """Return the canvas width — the viewBox ``width``."""
        return cast("int | float", self._data["width"])

    def height(self) -> int | float:
        """Return the canvas height — the viewBox ``height``."""
        return cast("int | float", self._data["height"])

    def elements(self) -> list[Element]:
        """Return the top-level elements, lazily wrapped on first access."""
        if self._elements is None:
            self._elements = [Element(el) for el in self._data["elements"]]

        return self._elements

File: src/csharp/core/src/Style/Canvas.cs
using System;
using System.Collections.Generic;
using System.Text.Json.Nodes;

namespace DiceBear.Internal
{
    /// <summary>
    /// Read-only view over a style definition's <c>canvas</c> block, exposing
    /// the drawing area dimensions and the top-level element list.
    /// </summary>
    internal sealed class Canvas
    {
        private readonly JsonObject _data;

        private IReadOnlyList<Element>? _elements;

        internal Canvas(JsonObject data) => _data = data;

        /// <summary>
        /// Returns the canvas width — the <c>width</c> value of the SVG
        /// <c>viewBox</c>.
        /// </summary>
        internal double Width() => JsonRead.Num(_data, "width") ?? 0.0;

        /// <summary>
        /// Returns the canvas height — the <c>height</c> value of the SVG
        /// <c>viewBox</c>.
        /// </summary>
        internal double Height() => JsonRead.Num(_data, "height") ?? 0.0;

        /// <summary>
        /// Returns the top-level elements rendered onto the canvas, lazily
        /// wrapped as <see cref="Element"/> instances on first access.
        /// </summary>
        internal IReadOnlyList<Element> Elements()
        {
            if (_elements is not null)
            {
                return _elements;
            }

            var array = JsonRead.Arr(_data, "elements");

            if (array is null)
            {
                return _elements = Array.Empty<Element>();
            }

            var result = new List<Element>(array.Count);

            foreach (var element in array)
            {
                if (element is JsonObject obj)
                {
                    result.Add(new Element(obj));
                }
            }

            return _elements = result;
        }
    }
}
